### PR TITLE
minor KTO setting changes + KL batch size

### PR DESCRIPTION
**Pull request by kawine**
# What does this PR do?
 
- change the default learning rate from 5e-7 to 1e-6; for models 7B and up (which is what most people use w/ or w/o LoRA), i this works better with the default beta=0.1 than an lr=5e-7, which can lead to slow learning
- add explanations in the docs on choosing the learning rate
- (important) change the batch size of the KL data to be the per-step batch size, not the effective batch size; the current implementation leads to worse results when gradient accumulation is used
- allow dropout to be used, like in DPOTrainer

cc @kashif @qgallouedec

### CODE REVIEW

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
in the `kto_config.py` docstrings can you kindly add:

```py
disable_dropout (`bool`, *optional*, defaults to `True`):
            Whether to disable dropout in the model.
```

**Comment by kawine** [comment]
> in the `kto_config.py` docstrings can you kindly add:
> 
> ```python
> disable_dropout (`bool`, *optional*, defaults to `True`):
>             Whether to disable dropout in the model.
> ```

done!

File: trl/trainer/kto_trainer.py [hunk 1]
```diff
@@ -73,7 +73,11 @@
 
 
 def _get_kl_dataset(batch: Dict[str, List[Any]]) -> Dict[str, List[Any]]:
-    """Creates mismatched pairs of prompts and completions for the KL dataset by adding a +1 offset to the order of completions."""
+    """
+    Creates mismatched pairs of prompts and completions for the KL dataset by adding a +1 offset to the order of completions.
+    For best results, mismatched outputs y' used to estimate the KL term for a batch should be the same set as the matched
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:78]:
```suggestion
    For best results, the mismatched outputs y' used to estimate the KL term for a batch should be the same set as the matched
```

File: docs/source/kto_trainer.mdx [hunk 2]
```diff
@@ -61,13 +63,17 @@ The KTO trainer expects a model of `AutoModelForCausalLM`, compared to PPO that
...
 By default, they are both 1. However, if you have more of one or the other, then you should upweight the less common type such that the ratio of (`desirable_weight` \\(\times\\) number of positives) to (`undesirable_weight` \\(\times\\) number of negatives) is in the range 1:1 to 4:3.
 
 <Tip>
-It is strongly recommended you use a learning rate between `5e-7` and `5e-6` with an effective batch size between `8` and `32`, for both LoRA and full finetuning. Even if you are working with a small dataset, we do not recommend using a learning rate outside this range; instead, using smaller batch sizes and/or more training epochs will give you better results.
+Every choice of `beta` has a maximum learning rate it will tolerate before learning degenerates. For the default `beta = 0.1', this learning rate is `1e-6` for most models. The lower beta is, the lower your learning rate should be. In general, we strongly recommend a learning rate between `5e-7` and `5e-6`. Even if you are working with a small dataset, we do not recommend using a learning rate outside this range; instead, use more epochs.
```
Comment (kashif) [inline docs/source/kto_trainer.mdx:72]:
```suggestion
Every choice of `beta` has a maximum learning rate it will tolerate before learning degenerates. For the default `beta = 0.1', this learning rate is `1e-6` for most models. The lower the beta is, the lower your learning rate should be. In general, we strongly recommend a learning rate between `5e-7` and `5e-6`. Even if you are working with a small dataset, we do not recommend using a learning rate outside this range; instead, use more epochs. 
```

**Review by kashif — APPROVED** [review]
